src/sparseml/pytorch/sparsification/pruning/modifier_powerpropagation.py

```diff
@@ -0,0 +1,327 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Modifier for models through powerproagation.
+
+"""
+
+
+import logging
+from typing import List, Optional, Union
+
+import torch
+from torch.nn import Conv2d, Linear, Module, Parameter
+from torch.nn import functional as F
+from torch.optim.optimizer import Optimizer
+
+from sparseml.optim import ModifierProp
+from sparseml.pytorch.sparsification.modifier import (
+    PyTorchModifierYAML,
+    ScheduledModifier,
+)
+from sparseml.pytorch.utils import (
+    NamedLayerParam,
+    get_named_layers_and_params_by_regex,
+    get_prunable_layers,
+    tensors_to_device,
+)
+from sparseml.pytorch.utils.logger import LoggerManager
+from sparseml.utils import ALL_PRUNABLE_TOKEN, ALL_TOKEN, validate_str_iterable
+
+
+__all__ = [
+    "PowerpropagationModifier",
+    "PowerpropagationWrapper",
+]
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class PowerpropagationWrapper(Module):
+    def __init__(self, layer: Module, alpha: float = 1.0):
+        super(PowerpropagationWrapper, self).__init__()
+
+        if not isinstance(layer, Conv2d) and not isinstance(layer, Linear):
+            raise ValueError("Powerpropagation only works with Linear and Conv layers")
+
+        self.layer = layer
+        # First set alpha to 1, then update it to the correct
+        # value. This avoids replicating the code that updates
+        # the layer weights.
+        self.register_buffer("alpha", torch.tensor(1.0, requires_grad=False))
+        self.set_alpha(alpha)
+
+    def forward(self, x):
+        weight = self.layer.weight * pow(abs(self.layer.weight), self.alpha - 1)
+
+        if isinstance(self.layer, Conv2d):
+            return F.conv2d(
+                x,
+                weight,
+                self.layer.bias,
+                self.layer.stride,
+                self.layer.padding,
+                self.layer.dilation,
+                self.layer.groups,
+            )
+        elif isinstance(self.layer, Linear):
+            return F.linear(x, weight, self.layer.bias)
+        else:
+            raise ValueError(
+                "Powerpropagation only works with Linear and Conv2d layers"
+            )
+
+    def set_alpha(self, new_alpha):
+        with torch.no_grad():
+
+            self.layer.weight *= pow(abs(self.layer.weight), self.alpha / new_alpha - 1)
```

Comment:
I didn't quite understand this part. From my understanding, this is the initializing the weight parameters. Doesn't the paper (page 3, second last para) mention this as: `\phi_i=sign(\theta_i) * pow(\abs{\theta_i}, \frac{1}{\alpha})`.

Reply:
This is a generalization of the formula you quote. So the formula you wrote is for updating the exponent when the current exponent is 1 (this is the `\frac{1}`). So replace the 1 with some other exponent, say `\rho`:
```
\phi_i=sign(\theta_i)* pow(\abs{\theta_i}, \frac{\rho}{\alpha})
= \theta_i*pow(\abs{\theta_i}, \frac{\rho}{\alpha}-1)
```

So this is what I have, except `\theta_i` is `self.layer.weight`, `\alpha` is `new_alpha`, and `\rho` is `self.alpha`.

Reply:
Yeah, makes sense. Thank you for clarifying.